src/KineticConvolution/Direction.cs

```diff
@@ -31,12 +31,132 @@ public Direction(IAlgebraicNumber x, IAlgebraicNumber y)
 
         public IAlgebraicNumber Y { get; }
 
-        public bool Equals(Direction d)
+        /// <inheritdoc />
+        public bool Equals(Direction? other)
         {
-            return
-                DirectionHelper.Determinant(this, d).IsZero() &&
-                X.Sign() == d.X.Sign() &&
-                Y.Sign() == d.Y.Sign();
+            if (other is null)
+            {
+                return false;
+            }
+
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
+
+            return DirectionHelper.Determinant(this, other).IsZero()
+                && X.Sign() == other.X.Sign()
+                && Y.Sign() == other.Y.Sign();
```

Comment:
If it is a class then the types should also be tested:

```suggestion
           if(GetType() != other.GetType())
           {
               return false;
           }
            return DirectionHelper.Determinant(this, other).IsZero()
                && X.Sign() == other.X.Sign()
                && Y.Sign() == other.Y.Sign();
```